### ISSUE TITLE
hook_update_N not run when upgrading from D7 for modules with low schema version

### ISSUE THREAD

**Issue by argiepiano** [backdrop/backdrop-issues#6055]
## Description of the bug

When upgrading a D7 site that uses modules with schema versions lower than 7000, Backdrop will not run any of the `hook_update_N` hooks for that module 

* Please read below about situations when a D7 module may have a 6XXX schema version. It's rare but possible!

## Steps To Reproduce

I have created some small custom modules to reproduce this:
- [testing_low_version-DRUPAL7.zip](https://github.com/backdrop/backdrop-issues/files/11213250/testing_low_version-DRUPAL7.zip)
- [testing_low_version-BACKDROP.zip](https://github.com/backdrop/backdrop-issues/files/11213254/testing_low_version-BACKDROP.zip)

To facilitate things, I'm also attaching a D7 database dump (vanilla install, only with the above module installed)
- [tenutowe_drupaltesting_2023-04-12.sql.gz](https://github.com/backdrop/backdrop-issues/files/11213269/tenutowe_drupaltesting_2023-04-12.sql.gz)

To reproduce:
1. Follow all the  normal preliminary steps to do a D7 upgrade (that is, put the Backdrop codebase in a folder, set up the database credentials, etc). Do not install your Backdrop site
2. Move the above BACKDROP module to the `modules` folder
3. Import the above D7 database into the Backdrop database
4. Run `/core/update.php` to create the update
5. Log into the new Backdrop site
6. Visit `/get-schema-version`

The above will output
<img width="879" alt="Screen Shot 2023-04-12 at 9 41 09 AM" src="https://user-images.githubusercontent.com/9938978/231510121-7f3725ff-b513-4404-9f01-7baba29393c0.png">

Three problems:
- The module is actually enabled in Backdrop (despite the low 6XXX schema version in the database)
- Schema version for the custom module is STILL 6066 after upgrade
- More seriously, the function `testing_low_version_update_1000()` NEVER runs. No other `hook_update_N()` is ever run oduring upgrade

## Actual behavior

The update hook is not run.

## Expected behavior

The update hook should run.

## Additional information

- Backdrop CMS version: 1.24.0

When does a Drupal 7 module have a 6XXX schema version?
-----

It's rare, but it can happen. All these conditions must exist:
1. The D7 module was original created for D6, AND
2. The D7 module implements `hook_update_last_removed()` returning a 6XXX number, AND
3. The D7 module does NOT implement `hook_update_N()`

There may be legitimate situations when you don't want run update_N for modules with 6XXX schema version (there is no upgrade path from D6 -> Backdrop). However, as demonstrated above, there are cases when a D7 module WILL have a 6XXX schema version.  

It's problematic that this module is enabled, but its update hooks are not run during upgrade.

Possible solutions
----
When the schema version is less than 7000: Either do not enable the offending module, or run the updates, or throw an error/warning.

**Comment by klonos** [comment]
@argiepiano this is not to say that we shouldn't try to fix this if we can, but if I am reading correctly the conditions you have outlined, then that means that the D6 db was not updated to the latest possible before updating to D7, which caused the db to have a 6xxx schema version in the system table for that module. Right? I know that our update instructions state the following (in [Prepare the Drupal site for an upgrade to Backdrop CMS](https://docs.backdropcms.org/documentation/step-2-preparation)):

> 2. **Update all modules and themes**:
>    Make sure all modules and themes are updated to their most recent version.

So this seems like a situation where things should be fixed in the D7 site, before attempting to upgrade to Backdrop. Right, or am I missing something?

**Comment by argiepiano** [comment]
> that means that the D6 db was not updated to the latest possible before updating to D7, which caused the db to have a 6xxx schema version in the system table for that module. Right?

That is not correct. The database is a Drupal 7 database. Try installing my small module in the vanilla D7 installation, then check the `system` table and look for the entry for that module. You'll see 6066 in the schema version. So, the database is a "D7 database" yet the schema version for the module is still a 6XXX version.

**Comment by klonos** [comment]
Perhaps I wasn't clear in what I was trying to say. Take two: if the 7.x module did not make sure that its schema had a 7xxx number, then things should be fixed in the D7 site **before** attempting to upgrade to Backdrop. So I guess what I'm saying is that the 7.x module should have a `hook_update_last_removed()` implementation with the greatest 6xxx number, and if that is not enough to remove the 6xxx number from the system table of the D7 db, then also even a "blank" `hook_update_7000()` that returns nothing (only adds the 7000 in the schema).

**Comment by argiepiano** [comment]
OK, got it. Thanks for the clarification. 

In an ideal world,  the D7 module maintainer would take care of this. However, I don't think this (creating an empty update_7000 hook just to be sure the schema version in the DB is 7000) is documented at all in Drupal's world. And at this point, given that D7 module maintenance is pretty null for many modules, it would be unlikely that a maintainer would merge and release a new version with that empty hook just so that we can upgrade the site to Backdrop.

The good news is that this is rare. I've seen this in `l10n_community`, which I ported (but haven't yet found a solution for people wanting to upgrade from D7->Backdrop and run into this problem), and `admin_menu_toolbar` (which is unlikely to be ported).

**Comment by klonos** [comment]
> In an ideal world, the D7 module maintainer would take care of this. However,...

Fair enough 👍🏼 

> The good news is that this is rare.

If all it would take to take care of this on our end would be a few lines of simple code changes, then I wouldn't be against it.

**Comment by kiamlaluno** [comment]
Drupal 7 modules are not required to update their schema version. The only way to achieve that is implementing a new `hook_update_N()` hook, but Drupal 7 modules are not required to have an empty update hook just to update their schema version.

Somebody could argue that Drupal 7 modules that have not needed to implement an update hook for Drupal 7 for all the Drupal 7 lifetime are a minority or very simple modules. Still, assuming that the schema version for a Drupal 7 module is a number higher than 6999 is wrong.

**Comment by kiamlaluno** [comment]
Notice also that, in the case described in this issue, Backdrop will never run any update hooks, even when the module would later implement a new update hook, because the schema version would keep its 6xyx value.

The only workaround is for the Backdrop module to override its own schema version in the *system* table.

**Comment by klonos** [comment]
> ...assuming that the schema version for a Drupal 7 module is a number higher than 6999 is wrong.

Agreed 👍🏼 ...we might however need to mimic that (or actually do it on behalf of such modules during the upgrade) in order to solve this issue here.

**Comment by kiamlaluno** [comment]
Assuming that Backdrop does not want to run Drupal 6 updates, the only way to fix this is letting Backdrop core change the schema version for a module to 7000 when it is lower than 7000. (I guess this should be the case, since Backdrop core does not implement the functions a Drupal 6 module would call.)
